src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/JsonRpc/JsonRpcRequestScheduler.cs

```diff
@@ -0,0 +1,112 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.VisualStudio.Threading;
+using OmniSharp.Extensions.JsonRpc;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.JsonRpc
+{
+    internal class JsonRpcRequestScheduler : IDisposable
+    {
+        private readonly Task _processingQueueTask;
+        private readonly AsyncQueue<QueueItem> _queue;
+        private readonly CancellationTokenSource _disposedCancellationTokenSource;
+        private readonly ILogger<JsonRpcRequestScheduler> _logger;
+
+        public JsonRpcRequestScheduler(ILoggerFactory loggerFactory)
+        {
+            _disposedCancellationTokenSource = new CancellationTokenSource();
+            _logger = loggerFactory.CreateLogger<JsonRpcRequestScheduler>();
+            _queue = new AsyncQueue<QueueItem>();
+
+            // Start the queue processing
+            _processingQueueTask = ProcessQueueAsync();
+        }
+
+        public bool Schedule(RequestProcessType type, string identifier, ProcessSchedulerDelegate processAsync)
+        {
+            lock (_disposedCancellationTokenSource)
+            {
+                if (_disposedCancellationTokenSource.IsCancellationRequested)
+                {
+                    // Shutting down
+                    return false;
+                }
+            }
+
+            var queueItem = new QueueItem(type, identifier, processAsync);
+
+            // Try and enqueue item, if this fails it means we're tearing down.
+            var enqueued = _queue.TryEnqueue(queueItem);
+            return enqueued;
+        }
+
+        public void Dispose()
```

Comment:
Drain the queue? Complete the queue? Something!?

The locking here seems odd to me, and I think completing the queue might help avoid it.. essentially, if you tell the queue to stop accepting new messages, then drain it if you like, you don't need to worry about `Schedule` and `Dispose` overlapping, right?

Reply:
> Drain the queue? Complete the queue? Something!?

Hahaha, so cancelling the disposed cancellation token effectively completes the queue. It results in the abort of the currently running request and then the queue itself fails to re-execute because cancellation is requested.

> The locking here seems odd to me, and I think completing the queue might help avoid it..

You're right, we don't really need to lock here. If more requests are scheduled during shutdown everything is tearing down anyways and they just wont complete. I'm going to remove that behavior because it's not valuable.

Oh and one thing to mention, we don't really need to drain the queue because our language server operates as  a 100% language server. Meaning, it gets disposed/torn down and requests don't actually need to run on shutdown because when another solution is opened we spin up a new queue / language server etc.